src/js/tooltip.js

```diff
@@ -0,0 +1,420 @@
+	 * @param {Object} opts - An options object for configuring the tooltip (Optional)
+	*/
+	constructor (tooltipEl, opts) {
+
+		if (!Tooltip._tooltips) {
+			Tooltip._tooltips = new Map();
+		}
+
+		this.tooltipEl = tooltipEl;
+		Tooltip._tooltips.set(this.tooltipEl, this);
+
+		this.opts = opts || Tooltip.getOptions(tooltipEl);
+		this.opts = Tooltip.checkOptions(this.opts);
+
+		this.target = new Tooltip.Target(document.getElementById(this.opts.target));
+		this.tooltipPosition = this.opts.position;
+		this.tooltipAlignment = null;
+		this.visible = false;
+
+		this.delegates = {
+			doc: new Delegate(),
+			tooltip: new Delegate(),
+		};
+
+		Viewport.listenTo('resize');
+
+		// Do you render as soon as possible?
+		if (this.opts.renderOnConstruction) {
+			this.render();
+			this.show();
+		}
+	};
+
+	/**
+	 * Get the data attributes from the tooltipEl. If the tooltip is being set up
+	 * declaratively, this method is used to extract the data attributes from
+	 * the DOM.
+	 * @param {HTMLElement} tooltipEl - The tooltip element in the DOM (Required)
+	 * @todo - refactor this out to smartly iterate over data attributes
+	*/
+	static getOptions(tooltipEl) {
+		let opts = {};
+		if (tooltipEl.hasAttribute('data-o-tooltip-position')) {
+			opts.position = tooltipEl.getAttribute('data-o-tooltip-position');
+		}
+		if (tooltipEl.hasAttribute('data-o-tooltip-target')) {
+			opts.target = tooltipEl.getAttribute('data-o-tooltip-target');
+		}
+
+		if (tooltipEl.hasAttribute('data-o-tooltip-render-on-construction')){
+			opts.renderOnConstruction = (tooltipEl.getAttribute('data-o-tooltip-render-on-construction') === 'true');
+		}
+
+		if (tooltipEl.hasAttribute('data-o-tooltip-z-index')){
+			opts.zIndex = tooltipEl.getAttribute('data-o-tooltip-z-index');
+		}
+		return opts;
+	};
+
+	/**
+	 * Check the options passed in are valid, and that the required option
+	 * (target) is present
+	 * @param {Object} opts - An Object with configuration options for the tooltip
+	 * @throws o-tooltip error: opts.target us not set
+	 * @throws o-tooltip error: opts.tooltipPosition is not one of "above", "below"
+	 * "left" or "right"
+	*/
+		static checkOptions(opts) {
+
+		if(!opts.target) {
+			Tooltip.throwError("tooltip.target is not set. An target for the tooltip to point at must be provided");
+		}
+
+		// Check that the value of tooltip position is valid. Default to below.
+		if (opts.tooltipPosition) {
+			if (Tooltip.validTooltipPositions.indexOf(opts.tooltipPosition) === -1) {
+				Tooltip.throwError("Invalid value for tooltip position. Valid values are:" + Tooltip.validTooltipPositions.toString() +" or nothing which will default to a value of `below`");
+			}
+		} else {
+			opts.tooltipPosition = "below";
+		}
+
+		return opts;
+	};
+
+	/**
+	 * Render the tooltip. Adds markup and attributes to this.tooltipEl in the DOM
+	*/
+	render() {
+
+		this.tooltipEl.setAttribute('role', 'tooltip');
```

Comment:
I think I might be misunderstanding something. If `render` is not called it looks like the `role` is not added to the tooltip element. All the demos have `this.opts.renderOnConstruction` set to `true` which means we have no demo showing what it is like when the tooltip is not rendered upon construction.